Polar 2.0

etpinard opened the pull request:
should close everything listed under the on-par polar [milestone](https://github.com/plotly/plotly.js/milestone/2) :tada: 

--------------

I'm making a PR at this stage to:

- hopefully get a few :+1: s about the attributes came up with. I would appreciate if @alexcjohnson @cldougl @chriddyp could take a :eyes: at commits https://github.com/plotly/plotly.js/commit/3c8ae5beb88b32ab2158255a15eca5157fbb5b4e and https://github.com/plotly/plotly.js/commit/853581e3ede84aa412d17fdb3b5d43e2177f63ae,
- pin down the last open items, and
- set a firm base branch for @dfcreative 's upcoming `scatterpolargl` work.

### Working features so for:

- `scatterpolar` is on-par with `scattertenary` e.g. lines, markers, text, some fills options and line shapes as well as point selections are supported.
- `scatterpolar` accepts input angles in both degrees and radians using a polar-only `thetaunit` attribute
- `polar` subplots have fully configurable radial and angular axes styling.
- users can set the direction and starting angle of each `polar` subplots. (see [`polar_direction`](https://github.com/plotly/plotly.js/blob/543b61531073bb3f7f81157b8d28e56c2b495f46/test/image/baselines/polar_direction.png))
- `polar` subplots can be carve into sectors (see [`polar_sector`](https://github.com/plotly/plotly.js/blob/543b61531073bb3f7f81157b8d28e56c2b495f46/test/image/baselines/polar_sector.png))
- `polar` subplots can display angular tick labels in degrees or in fractions of pi.
- `polar` subplots accepts category as angles (see [`polar_categories`](https://github.com/plotly/plotly.js/blob/543b61531073bb3f7f81157b8d28e56c2b495f46/test/image/baselines/polar_categories.png))

Quick demo:

![peek 2017-12-12 09-40](https://user-images.githubusercontent.com/6675409/33890192-7b9838d4-df20-11e7-86b6-3e5c08f25ee4.gif)

### Biggest open item: zoombox interactions

What should zoom interactions look like? At first, me and @alexcjohnson thought that a viewbox-based zoom system would suffice. In details, we thought about exposing a set of scalar attributes e.g. `polar.x`, `polar.y` and `polar.zoom` that determine the location and zoom-level of the polar subplots with respect to their `polar.domain` bounding-box. But there's a problem with such system: for some sets of `x/y/zoom` values, viewers won't see the subplot radial and/or angular axes - making data point only discernible via hover. 

While adding functionality for non-full `polar.sector` values, I thought that perhaps we could make zoombox interactions correspond to a combination of polar `sector` and radial axis range relayout updates. That is, zooming in on a point would correspond to carving a sector around it and narrowing the `radialaxis.range` (note that radial axis ranges that don't start at 0 are common in other libraries e.g. [matlab](https://www.mathworks.com/help/matlab/ref/rlim.html)). 

@nicolaskruchten pointed out that generalising `polar.sector` to some 4-pt polar bounding box attribute could allow for even better zooming interactions (we should probably draw that on a whiteboard and take a screenshot to make it easier to grasp).

So I ask: is `x/y/zoom` zooming more intuitive than some fancy sector/radial range algo?

----

cc'ing @jackparmer @cpsievert @monfera @bpostlethwaite who might also be interested.

etpinard commented on src/plots/cartesian/axes.js:
```diff
@@ -819,6 +823,13 @@ axes.calcTicks = function calcTicks(ax) {
+    // If same angle over a full circle, the last tick vals is a duplicate.
+    //
+    // TODO must do something similar for angular date axes.
+    if(ax._id === 'angular' && Math.abs(rng[1] - rng[0]) === 360) {
```
I had to hack a few things deep into `axes.js` to get the angular axes looking the way I wanted. I hope @alexcjohnson won't mind.

alexcjohnson replied:
What if you use something else as the period of a linear angular axis? Also do we need to do something more flexible to handle *nearly* duplicate values so we don't get overlaps?

etpinard commented on src/plots/polar/layout_attributes.js:
```diff
@@ -0,0 +1,327 @@
+        //  http://www.seasonaladjustment.com/2012/09/05/clock-plot-visualising-seasonality-using-r-and-ggplot2-part-3/
+        //  https://i.pinimg.com/736x/49/b9/72/49b972ccb3206a1a6d6f870dac543280.jpg
+        //  https://www.climate-lab-book.ac.uk/spirals/
+    },
```
Some screenshot of date angular axes:

![image](https://user-images.githubusercontent.com/6675409/33892182-285e02ec-df26-11e7-92c4-71f864820b34.png)

![image](https://user-images.githubusercontent.com/6675409/33892200-32ec6104-df26-11e7-9215-d702616a1f29.png)

![image](https://user-images.githubusercontent.com/6675409/33892242-49039ef8-df26-11e7-875f-96618c712e43.png)


![image](https://user-images.githubusercontent.com/6675409/33892282-6bc194f4-df26-11e7-83aa-e365db28a934.png)

srpn replied:
Great feature !! Can you please include the code (layout config attributes) for the above example screenshots?

etpinard commented on src/plots/cartesian/axes.js:
```diff
@@ -1905,7 +1905,7 @@ axes.doTicks = function(gd, axid, skipTitle) {
         counterLetter = axes.counterLetter(axid),
         vals = axes.calcTicks(ax),
-        datafn = function(d) { return [d.text, d.x, ax.mirror].join('_'); },
+        datafn = function(d) { return [d.text, d.x, ax.mirror, d.font, d.fontSize, d.fontColor].join('_'); },
```
Cartesian axes use a special [`ticks`](https://github.com/plotly/plotly.js/blob/599dff93f9923e7ef4d6446de734cfb4eb39d663/src/plot_api/subroutines.js#L484) update pathway that.

This here gets `Axes.doTicks` to update properly. I'm pretty sure some axis style updates are currently broken for ternary subplots. I'll lock those down in tests soon.

etpinard replied:
:lock: in https://github.com/plotly/plotly.js/pull/2200/commits/7b0f718d8961794fc395ef249cdb6462c25914b9

chriddyp commented:
🎉 🏆

alexcjohnson commented:
Re: zooming interaction: I think we may need two different modes, eventually if not right away, because there are two very different ways polar plots get used. One is where you're drawing a real 2D space (where zero is a unique point and shapes have meaning) just in polar coordinates. That's where you would want zoombox-type interactions that I was thinking of. I notice though that none of your examples are of this type (except maybe the cardioid). Your examples all have independent R and Theta variables, just mapped onto a polar space mostly to show periodicity. Incidentally, this points to another way the behavior may want to depend on this distinction: what do you do with negative R? If you're mapping a real space, then it should show up as a positive R in the opposite direction, but if you're not it should disappear at the origin.

So how should zooming work in the independent-variable case? I still need to look at some more example plots with the question "how would I want to zoom this to explore THIS data set", but as a first cut, I feel like the most useful for people would be to independently move each of the axis ends - so min and max radius (dragging them like you drag an axis end in cartesian? @etpinard points out that this is ambiguous with interacting with data near the origin though... not sure how to handle that.), and for angle I guess initial angle (rotating the whole plot) and period (so points at the initial angle would stay put, points at initial angle + period would move with the cursor, points at initial angle + twice the period would move twice as fast as the cursor... per private discussion with @etpinard this interaction should NOT happen when the period is physically significant, like 360 degrees or 2Pi - perhaps we can have an attribute to enable/disable it, with a smart default set depending on whether an explicit period is provided or not)?

I also liked the way our old polar plots had the radial axis on top of the data, with an outline around the text in the background color. Perhaps it could be configurable whether this axis is above or below the data?

etpinard commented:
> I also liked the way our old polar plots had the radial axis on top of the data, with an outline around the text in the background color. Perhaps it could be configurable whether this axis is above or below the data?

Absolutely. I'll add a `polar.radialaxis.layer` enumerated attribute. I could even make that the default if you prefer. Moreover, it's true that polar charts would benefit a lot from https://github.com/plotly/plotly.js/issues/1597

chriddyp commented:
> What should zoom interactions look like?

In antenna design, where `r` represents a meaningful amplitude in DBs, it seems like its helpful to be able to zoom just `r` in and out, i.e. change the "`rmin`" value: 
Here is the example from matlab
![](https://www.mathworks.com/help/examples/antenna/win64/xxanimated.gif)

etpinard commented on src/traces/scatterpolar/plot.js:
```diff
@@ -10,33 +10,59 @@
+                    var rad = cdi.rad;
+                    cdi.x = rr * Math.cos(rad);
+                    cdi.y = rr * Math.sin(rad);
+                    continue;
```
cc @dfcreative

etpinard commented on src/plots/polar/constants.js:
```diff
@@ -15,15 +15,17 @@ module.exports = {
+        'angular-axis',
+        'radial-axis',
+        'angular-line',
+        'radial-line'
```
cc @alexcjohnson 's 

> I also liked the way our old polar plots had the radial axis on top of the data, with an outline around the text in the background color. Perhaps it could be configurable whether this axis is above or below the data?

from https://github.com/plotly/plotly.js/pull/2200#issuecomment-351127938

etpinard commented:
@alexcjohnson I just pushed about 20 commits where I:

- updated the attributes as requested in https://github.com/plotly/plotly.js/commit/3c8ae5beb88b32ab2158255a15eca5157fbb5b4e   and https://github.com/plotly/plotly.js/pull/2200/commits/853581e3ede84aa412d17fdb3b5d43e2177f63ae (where I just made a few comments), 
- added support for `radialaxis.title`, 
- merged master :sweat: and 
- fixed a few things while :palm_tree: a few other things

----

I'll spend a few hours tomorrow trying to get angular dates to work better with `angularaxis.period`. The main open item I believe is what to do with negative `r` values (referencing https://github.com/plotly/plotly.js/commit/3c8ae5beb88b32ab2158255a15eca5157fbb5b4e#r26416872). Other than that, this thing is pretty close.

alexcjohnson commented:
Currently if you zoom so some points are below the minimum radius, points disappear (as they should) but so do lines and fills (this is the polar_categories mock):
![screen shot 2018-01-04 at 8 17 32 pm](https://user-images.githubusercontent.com/2678795/34591564-7eea68f8-f18c-11e7-9c42-2950fd5c8ec3.png)
Seems like the lines and fills should continue into the center, right? That would be as analogous as we can get to how cartesian handles off-plot points.

alexcjohnson commented:
Radial axis dragging seems to be circling the wrong point:
![kapture 2018-01-04 at 21 46 18](https://user-images.githubusercontent.com/2678795/34593461-fc301324-f198-11e7-855a-6a2f5d3889fa.gif)
Also we need a cursor or something (`move`? `grab` with fallbacks for IE?) to show its grabber.

alexcjohnson commented:
I don't see where you're generating fractions of Pi, but at some precision (denominator >= 100?) it should switch to decimal - decimal fractions of Pi? Also should hovertext have these Pi fractions/decimals too instead of raw decimal numbers?
![screen shot 2018-01-04 at 10 29 16 pm](https://user-images.githubusercontent.com/2678795/34594422-ef538cf2-f19e-11e7-8198-f2786e237220.png)

alexcjohnson commented:
I find the radial zoom effect a lot of fun but a bit confusing to operate. I always know what's going to happen when I mouse up, that part is great - but I don't always know where to put my cursor to get what I want. It sometimes swaps between keeping the donut, keeping the hole, and keeping the whole thing, and it's not obvious to me when and why it's going to change. Might help to have some kind of handle for the starting point, where you moused down? And possibly some sort of guides delineating regions with different results, or just a clearer rule about which is which? 

Perhaps it could be as simple as: take the ray from the center toward where you moused down; the 180 degree sector clockwise from that ray gives you the donut between mouse down and mouse up, the sector ccw from that ray gives you a new outer radius only (set by where you mouse up, ignoring where you moused down).

It's also not obvious to me how to cancel the zoom in case I discover I've started it badly or something - which in cartesian I can do by making the box very small so it disappears. Sometimes I can't find any place to put the cursor to do this. Perhaps if there was a starting handle, I would be tempted to do the same thing as in cartesian, bring my cursor close to that point and the box would disappear?

alexcjohnson commented on src/plots/cartesian/axes.js:
```diff
@@ -974,10 +993,15 @@ axes.autoTicks = function(ax, roughDTick) {
     }
+    else if(ax._id === 'angular') {
+        ax.tick0 = 0;
+        base = getBase(1);
```
what am I missing? Isn't `getBase(1)` just 1^something === 1?

etpinard replied:
good catch -> 7f7193d

alexcjohnson commented on src/plots/cartesian/axes.js:
```diff
@@ -1399,6 +1424,66 @@ function formatLinear(ax, out, hover, extraPrecision, hideexp) {
     out.text = numFormat(out.x, ax, hideexp, extraPrecision);
 }
 
+function formatAngle(ax, out, hover, extraPrecision, hideexp) {
```
Ah nice, here it is 🎉  I guess even though it's currently only used for polar, there *is* something nice about having this in `axes.js` (granted that perhaps all of this should go somewhere else outside cartesian at some point) because I could imagine us wanting to support fractions and/or multiples of π on cartesian axes at some point.

But yeah, previous comments hold: we need a cutoff to switch to decimal, and then perhaps hover can use π too!

alexcjohnson commented:
It wasn't obvious to me at first that `pan` mode meant "rotate the plot" - I just stumbled on rotating and wondered why I couldn't zoom anymore. Maybe that's OK, but alternatively, perhaps you could grab in the ring of the angular axis labels to rotate, without going into a separate pan mode? Also it would be neat if it would snap to straight orientations, ie with zero angle at N, S, E, or W.

etpinard commented:
@alexcjohnson I think all your comments from last week (starting after https://github.com/plotly/plotly.js/pull/2200#issuecomment-355406922) have been addressed. I didn't push the improvements I've made to drag interactions to this branch yet (I'm looking for some feedback first :wink: ), they are [here](https://github.com/plotly/plotly.js/compare/polar...polar-dev).

Here's the new angular drag layer in action:

![peek 2018-01-11 18-24](https://user-images.githubusercontent.com/6675409/34852461-c2959b66-f6fc-11e7-9e6d-550dfae5befe.gif)

where I kept the `crosshair` cursor on the radial drag layer, and made the angular layer have `move`, to not confused them as they partially overlap.

------------

Concerning `dragmode: 'zoom'`, I added a few ways to _cancel_ the zoom event:

for example, if we did reach min zoom for two-sided re-range level and then moved back in the opposite direction reaching the mouse down radial distance, the zoom event is cancelled:

![peek 2018-01-11 18-27](https://user-images.githubusercontent.com/6675409/34852523-21e94248-f6fd-11e7-887e-bb334361efb4.gif)

if we did cross the origin, we get:

![peek 2018-01-11 18-30](https://user-images.githubusercontent.com/6675409/34852615-8c2acb0e-f6fd-11e7-9e88-62458830d0f2.gif)



Other than that, `dragmode: 'zoom'` behaves the same as previously. I hoping to add way to (easily) cancel zoom events without over-complicating things. I hope you'll like the results.

alexcjohnson commented:
Nice! I'll check out the other changes later tonight, but re: drag interactions:
- Angular - love it. Only addition I'd like at some point is snap to NSEW (for both the angular orientation and radial line orientation) but that can go in a TODO issue.
- Radial - definite improvement but I still find it confusing. I don't love the statefulness, ie if you start dragging inward it behaves one way, outward a different way, crossing the origin changes what you can do later... I feel like we can find a solution that only depends on the start and end points of the drag. Also wonder if we can get away from all angular dependence and only care about initial and final radii? It just always seems like angular dependence will give weird invisible boundaries with unexpected effects when you cross them. Sorry to be a pain about this one, it's a hard problem I don't have real clear answers on, just kind of thinking out loud and I'm probably going to contradict some suggestions I made earlier:

There are four radial range change you might want: zoom to a range, give just a new outer bound, give just a new inner bound, or abort.
- Abort: can we just say `Math.abs(r1 - r0) < MINDRAG` means abort?
  - The limitation this imposes is that you can't zoom straight to a tiny slice of the original radial range. But you can always zoom multiple times if that's what you want, and that's the same limitation we have with cartesian zoom.
- New outer bound: how about starting OR ending the drag within a certain region around the center (perhaps again `MINDRAG` px? Or that might be too small, we might want `2 * MINDRAG`)? So then you could either drag from the center outward to your new outer bound, or start at your new outer bound and drag to the center.
  - The limitation this imposes is that you can't make a very small tweak to the minimum bound, because it'll snap to the existing minimum. That doesn't strike me as a big deal though, it's a fairly weird case. And once we have `hole` as an available option (which is mostly when you'd want a variable inner bound anyhow!), we can remove snapping for the drag endpoint because any endpoint inside the hole counts as "no change to the minimum". So you can drag from the outside inward to an arbitrary number of pixels away from the edge of the hole without snapping to the edge. Seems though like the start of the drag will still need to snap to the hole edge, otherwise you can't ever hit the edge exactly.
- New inner bound: similar to new outer bound, either drag from the new edge you want all the way to (beyond) the outer edge, or drag from near the outer edge (snap to the edge) to the new edge you want.
- Zoom to a range is anything not covered by the above: your starting and ending radii are the new range.

alexcjohnson commented on src/plots/polar/layout_attributes.js:
```diff
@@ -211,16 +210,15 @@ var angularAxisAttrs = {
-            'corresponds to a line pointing right (like what mathematicians prefer).',
-            'For example to make the angular axis start from the North (like on a compass),',
-            'set `angularaxis.rotation` to *90*.'
+            'By default, polar subplots will `direction` set to *counterclockwise*',
```
~~will~~ with

alexcjohnson commented on src/plots/polar/layout_defaults.js:
```diff
@@ -113,6 +107,26 @@ function handleDefaults(contIn, contOut, coerce, opts) {
+                // Here, if a date angular axes is detected, we make
+                // all its corresponding traces invisible, so that
+                // when we do add support for data angular axes, the new
+                // behavior won't conflict with existing behavior
```
Beautifully done. Where should we document the use of categories for displaying polar dates? Just in an example somewhere? I'm thinking that doing it really right would involve categories like "Jan 2010", "Feb 2010", ... "Jan 2011" etc., `angularaxis.period` and then `tickvals`/`ticktext` to get labels to just be "Jan", "Feb", "Mar"...

etpinard replied:
In https://github.com/plotly/plotly.js/pull/2200/commits/8b4c2b5bd9e4303599f3a407a7261bb4c3a31708, I restricted the period for category axes to be `>=` than the number of categories, as I couldn't think of a use case for winding category labels. I don't think they make sense **unless** the users set angular `tickvals`/`ticktext`. That might not be too much to ask though :thinking: 

So, we could allow `period < ax._categories.length` to make date-like categories work with a period less than the data range or simply wait for a true official handling of angular dates. I'm not sure it's worth the trouble, but I'll leave it up to you to decide.

alexcjohnson replied:
I wouldn't *prohibit* `period < ax._categories.length` - someone will find a use for it for sure. As long as ticks only display for the first revolution, ideally with the ability to simplify them with `tickvals`/`ticktext`, and hovertext displays the actual angular data value so you know which revolution you've hovered on. But it's also a weird case so I'm fine with leaving this feature in a polar TODO issue.

alexcjohnson commented on test/jasmine/tests/polar_test.js:
```diff
@@ -232,8 +232,21 @@ describe('Test relayout on polar subplots:', function() {
+        .then(function() {
+            expect(gd._fullLayout.polar._subplot.viewInitial['radialaxis.range'])
+                .toBeCloseToArray([0, 11.225]);
+        })
```
Good catch! Just for my peace of mind so I don't have to trust that `viewInitial` does the right thing, can you also `expect` that `fullLayout.polar.radialaxis.range` matches `viewInitial[...range]`?

etpinard replied:
Good call. Fixed in https://github.com/plotly/plotly.js/pull/2200/commits/473643cf8dfb12f1543bd8dcb01ca00c568f6398

alexcjohnson commented on src/traces/scatter/line_points.js:
```diff
@@ -60,9 +60,10 @@ module.exports = function linePoints(d, opts) {
+        var di = d[index];
+        var x = xa.c2p(di.x);
+        var y = ya.c2p(di.y);
+        if(x === BADNUM || y === BADNUM) return di.intoCenter || false;
```
Very clean solution, love it!

alexcjohnson commented on test/image/mocks/polar_fills.json:
```diff
@@ -13,13 +13,37 @@
+  }, {
+    "type": "scatterpolar",
+    "mode": "lines",
+    "r": [5, 1, 2, 4, 5],
```
This test doesn't show that lines go to the center while markers are not displayed there (though I've seen in playing with it that you indeed got it right 🎉 ) - can you just show markers here, and bump the r=2 value down so no marker *should* show at the center?

etpinard replied:
See https://github.com/plotly/plotly.js/pull/2200/commits/473643cf8dfb12f1543bd8dcb01ca00c568f6398
for what I came up with. I'm not 100% sure this is what you meant though.

alexcjohnson replied:
Yep, you got it, thanks. I was just looking to see a corner in the line at the origin, with no corresponding marker. Lo and behold, there it is!
<img width="175" alt="screen shot 2018-01-12 at 5 53 38 pm" src="https://user-images.githubusercontent.com/2678795/34900931-9326709c-f7c1-11e7-9818-512bfc1bcc05.png">

alexcjohnson commented on src/plot_api/plot_api.js:
```diff
@@ -1946,7 +1946,7 @@ function _relayout(gd, aobj) {
                 // clear polar view initial stash for radial range so that
                 // value get recomputed in correct units
-                if(fullLayout._subplots.polar && fullLayout[p.parts[0]]) {
+                if(Array.isArray(fullLayout._subplots.polar) && fullLayout._subplots.polar.length && fullLayout[p.parts[0]]) {
```
Oh, I didn't notice this logic before - should it be further restricted to when the axis (I guess this is `p.parts[1]`) is `radialaxis`?

etpinard replied:
Good call -> https://github.com/plotly/plotly.js/pull/2200/commits/473643cf8dfb12f1543bd8dcb01ca00c568f6398

alexcjohnson commented on src/plots/polar/polar.js:
```diff
@@ -807,6 +788,23 @@ proto.updateMainDrag = function(fullLayout, polarLayout) {
+            Plotly.relayout(gd, updateObj);
+        }
+
+        Fx.click(gd, evt, _this.id);
```
Hmm, this is not specific to polar, you're just replicating what we have elsewhere... but it occurs to me, having the `doubleClick` handled before `click` (in `Fx.click`) is both different from native events (which always seem to put the `doubleClick` last, though some browsers - [*cough* IE *cough*](https://stackoverflow.com/questions/5497073/how-to-differentiate-single-click-event-and-double-click-event) - don't emit the second `click` at all) AND potentially incorrect: if `Plotly.relayout` is synchronous, then `Fx.click` will make it look like you clicked on whatever point happens to be under your cursor after autoranging.

etpinard replied:
There's https://github.com/plotly/plotly.js/issues/1546 that discuss the subject where @monfera argued for a new event called `plotly_singleclick` to go around this problem. Not sure what's best.

alexcjohnson replied:
Ah thanks for bringing up that issue - I'll just add this comment there.

alexcjohnson commented on test/jasmine/tests/polar_test.js:
```diff
@@ -496,3 +499,153 @@ describe('Test relayout on polar subplots:', function() {
+            }, 'after doubleclick');
+        })
+        .catch(fail)
+        .then(done);
```
Great test, very readable!

Presumably polar also handles right-clicks the same way as cartesian? Want to 🔒 it ala https://github.com/plotly/plotly.js/blob/master/test/jasmine/tests/click_test.js#L173-L213?

etpinard replied:
done in https://github.com/plotly/plotly.js/pull/2200/commits/473643cf8dfb12f1543bd8dcb01ca00c568f6398

etpinard commented:
Another day, another polar zoombox attempt, see https://github.com/plotly/plotly.js/compare/polar...polar-dev

@alexcjohnson First off, thanks very much for the long writeup in https://github.com/plotly/plotly.js/pull/2200#issuecomment-357117251, I think this is by far the best version so far. This gif should cover all the cases:

![peek 2018-01-12 14-18](https://user-images.githubusercontent.com/6675409/34891470-ba46bdf8-f7a3-11e7-860b-40809d71ab48.gif)

alexcjohnson commented:
> Another day, another polar zoombox attempt, I think this is by far the best version so far.

Love it, I think we have a winner! 🏆 Particularly nice how if you start at one end, there's no snap to the other end - because, really, why would you do that, it would be a noop. That gets around my concern about not being able to make a small change to the inner radius. Very very nice.

I could imagine decreasing the snap length a little bit, at least for small plots. But we can play with that later, the hard part is the mechanics and I think you've nailed it.

etpinard commented on src/plots/polar/constants.js:
```diff
@@ -33,6 +33,9 @@ module.exports = {
+    MINZOOM: 20,
+    // distance [px] off (r=0) or (r=radius) where we transition
+    // from single-sided to two-sided radial zoom
+    OFFEDGE: 20
```
per https://github.com/plotly/plotly.js/pull/2200#issuecomment-357393518

> I could imagine decreasing the snap length a little bit

I made this smaller than last Friday's version.

alexcjohnson replied:
> I made this smaller than last Friday's version.

Looks good. I wish there was a way to indicate to the user which behavior they're going to get, like we do with single- or double-ended arrows for the cartesian axis draggers... but I don't see any obvious choices among built-in cursors. I suppose we could do some super-complicated thing where we choose among all eight single-ended arrows depending on which direction you are from the center... but that might even be more confusing since it would make it look like those eight sectors had differences in behavior. Lets leave it as is for now, and see how it feels as it starts getting used.

alexcjohnson commented on test/jasmine/tests/scatterpolar_test.js:
```diff
@@ -31,14 +39,80 @@ describe('Test scatterpolar trace defaults:', function() {
+        nums: 'r: 4.022892\nθ: 2.239991',
+        name: 'Trial 3'
+    }, {
+        desc: 'on work on log radial axis',
```
~on work~ on log radial axis

alexcjohnson commented on src/plots/polar/polar.js:
```diff
@@ -1071,6 +1093,31 @@ function pathSectorClosed(r, sector) {
+
+    // make a small cut on full sectors so that the
+    // inner region isn't filled
+    if(isFullCircle(sector)) s1 -= 1e-3;
```
FWIW pies solved this by making [two arcs at each radius](https://github.com/plotly/plotly.js/blob/ef181daa3f18483af769cb776059b9709b17311b/src/traces/pie/plot.js#L216-L219) so the annulus really is complete.

*Probably* not going to cause any issues, it's just missing a fraction of a pixel, but I worry a little about numerical stability - it has to determine the center of the circle based on the very small vector from `p00` to `p01` and I'm not sure how much precision SVG uses.

alexcjohnson commented:
Two problems I just encountered with interaction effects, both while looking at the polar_direction mock:
- Rotating the plot by dragging the angular axis: seems like this is centered on the wrong point (though dragging the radial axis around has the correct center now)
- Dragging the radial axis around: *sometimes* after I've done this once, it doesn't seem to allow it anymore, I can only change the outer limit of the radius no matter which direction I start the drag. But sometimes it still works indefinitely... I can't really figure out the pattern.

alexcjohnson commented on src/plots/polar/polar.js:
```diff
@@ -911,9 +916,19 @@ proto.updateAngularDrag = function(fullLayout, polarLayout) {
-        _this.clipPaths.circle.select('path').attr('transform',
-            strTranslate(cxx, cyy) + strRotate(da)
-        );
+        // 'un-rotate' marker and text points
```
Oh man, that is such a satisfying effect 🍹

alexcjohnson commented:
@etpinard Fantastic work! Aside from the three quick comments above (at least I hope the interaction bugs are quick 😅 ) the only thing I'd like to see before merging is an issue collecting open items, so we can point people to that when questions arise, and use it to prioritize future additions.

alexcjohnson commented on src/components/dragelement/index.js:
```diff
@@ -76,6 +76,10 @@ dragElement.unhoverRaw = unhover.raw;
+ *      clampFn (optional, function(dx, dy) return [dx2, dy2])
+ *          Provide custom clamping function for small displacements.
+ *          By default, clamping is done using `minDrag` to x and y displacements
+ *          independently.
```
Nicely done. And it works great to solve the radial/angular drag bug! 🎯

alexcjohnson commented:
A hard-fought 💃  - excellent work.

etpinard commented:
Merging. 

For future development ideas and updates, please subscribe to https://github.com/plotly/plotly.js/issues/2255

etpinard commented:
Oh and please note that (r,t) _legacy_ polar charts are now deprecated! Please switch over to `scatterpolar`.
